In [1]:
import pandas as pd
# load data
df = pd.read_csv("PythonForDataScienceCodeData/auto_mpg.csv")

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [3]:
df.dropna(inplace=True)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model_year    392 non-null    int64  
 7   origin        392 non-null    object 
 8   name          392 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 30.6+ KB


Choose Predictors, Our Target is MPG
predictors are all columns except name
let x store predictors
let x store target

In [5]:
x = df.iloc[:,1:8] # chooses columns 1-7 - all except MPG and Name - Stores Predictor
y = df.iloc[:,0] # chooses columns 0 - MPG - Stores Target

In [6]:
x

cylinders  displacement  horsepower  weight  acceleration  model_year  \
0            8         307.0       130.0    3504          12.0          70   
1            8         350.0       165.0    3693          11.5          70   
2            8         318.0       150.0    3436          11.0          70   
3            8         304.0       150.0    3433          12.0          70   
4            8         302.0       140.0    3449          10.5          70   
..         ...           ...         ...     ...           ...         ...   
393          4         140.0        86.0    2790          15.6          82   
394          4          97.0        52.0    2130          24.6          82   
395          4         135.0        84.0    2295          11.6          82   
396          4         120.0        79.0    2625          18.6          82   
397          4         119.0        82.0    2720          19.4          82   

     origin  
0       usa  
1       usa  
2       usa  
3       usa  
4       usa  
..      ...  
393     usa  
394  europe  
395     usa  
396     usa  
397     usa  

[392 rows x 7 columns]

In [7]:
x = pd.get_dummies(x)
x
# the purpose of using the get_dummies is so the categorical variable "origin" can be converted to a usable form such as true/false

cylinders  displacement  horsepower  weight  acceleration  model_year  \
0            8         307.0       130.0    3504          12.0          70   
1            8         350.0       165.0    3693          11.5          70   
2            8         318.0       150.0    3436          11.0          70   
3            8         304.0       150.0    3433          12.0          70   
4            8         302.0       140.0    3449          10.5          70   
..         ...           ...         ...     ...           ...         ...   
393          4         140.0        86.0    2790          15.6          82   
394          4          97.0        52.0    2130          24.6          82   
395          4         135.0        84.0    2295          11.6          82   
396          4         120.0        79.0    2625          18.6          82   
397          4         119.0        82.0    2720          19.4          82   

     origin_europe  origin_japan  origin_usa  
0            False         False        True  
1            False         False        True  
2            False         False        True  
3            False         False        True  
4            False         False        True  
..             ...           ...         ...  
393          False         False        True  
394           True         False       False  
395          False         False        True  
396          False         False        True  
397          False         False        True  

[392 rows x 9 columns]

In [8]:
# now divide data into train and test for both x,y
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=18)
# we set random state to 0, which means this data is not reproducible
# test size is set to 20% which means 20% of the dataset will be testing.

In [9]:
from sklearn.preprocessing import StandardScaler

# Since the data are not in the same unit, measurements or scale, we must standardise them
# the mean of every column/variable becomes 0, with sd 1 after
scale = StandardScaler()
scale.fit_transform(x_train)
scale.transform(x_test);

In [10]:
from sklearn.linear_model import LinearRegression

# now a linear regression model can be fitted, can later be used to build the model
# y = B0 + B1*X1 + B2*X2 + _ _ _ _ + Bn*Xn - Y is the target, B0 is intercept and Bn are the coefficient to predictors xi
reg = LinearRegression()
reg.fit(x_train, y_train)
m = reg.coef_
c = reg.intercept_
m,c

(array([-0.54558713,  0.02654372, -0.02611905, -0.00655203,  0.02991416,
         0.81058575,  0.82116362,  1.20416269, -2.0253263 ]),
 -17.655565969692997)

(array([-0.54558713,  0.02654372, -0.02611905, -0.00655203,  0.02991416,
         0.81058575,  0.82116362,  1.20416269, -2.0253263 ]),
 -17.655565969692997)
 
The array that appears first are the coefficients and the the final value that gets printed individually is the intercept.

In [11]:
# now this model can be used against the training and testing data set
y_pred_train = reg.predict(x_train) # predicting against training data
y_pred_test = reg.predict(x_test) # predicting against test data

To evaluate the performance of the model, we can assess the r squared score for train and test data

In [12]:
from sklearn.metrics import r2_score

r2_score_train = r2_score(y_train, y_pred_train)
r2_score_train

0.8216701956964187

This shows an accuracy of 82.16%.

In [13]:
r2_score_test = r2_score(y_test, y_pred_test)
r2_score_test

0.8301439385772666

This shows an 83.10% accuracy.